flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java

```diff
@@ -96,332 +222,274 @@ public void initializeState(StateInitializationContext context) throws Exception
-							//   1) if close() was called on the operator then exit the while loop
-							//   2) if not wait 50 ms and try again to fetch a new split to read
-
-							if (currentSplit == null) {
-								if (this.shouldClose) {
-									isRunning = false;
-								} else {
-									checkpointLock.wait(50);
-								}
-								continue;
-							}
-						}
-
-						if (this.format instanceof CheckpointableInputFormat && currentSplit.getSplitState() != null) {
-							// recovering after a node failure with an input
-							// format that supports resetting the offset
-							((CheckpointableInputFormat<TimestampedFileInputSplit, Serializable>) this.format).
-								reopen(currentSplit, currentSplit.getSplitState());
-						} else {
-							// we either have a new split, or we recovered from a node
-							// failure but the input format does not support resetting the offset.
-							this.format.open(currentSplit);
-						}
-
-						// reset the restored state to null for the next iteration
-						this.currentSplit.resetSplitState();
-						this.isSplitOpen = true;
-					}
-
-					LOG.debug("Reading split: " + currentSplit);
-
-					try {
-						OT nextElement = serializer.createInstance();
-						while (!format.reachedEnd()) {
-							synchronized (checkpointLock) {
-								nextElement = format.nextRecord(nextElement);
-								if (nextElement != null) {
-									readerContext.collect(nextElement);
-								} else {
-									break;
-								}
-							}
-						}
-						completedSplitsCounter.inc();
-
-					} finally {
-						// close and prepare for the next iteration
-						synchronized (checkpointLock) {
-							this.format.close();
-							this.isSplitOpen = false;
-							this.currentSplit = null;
-						}
-					}
+		switch (state) {
+			case IDLE:
+				switchState(ReaderState.CLOSED);
+				break;
+			case CLOSED:
+				LOG.warn("operator is already closed, doing nothing");
+				return;
+			default:
+				switchState(ReaderState.CLOSING);
+				while (!state.isTerminal()) {
+					executor.yield();
 				}
+		}
 
-			} catch (Throwable e) {
-
-				getContainingTask().handleAsyncException("Caught exception when processing split: " + currentSplit, e);
+		try {
+			sourceContext.emitWatermark(Watermark.MAX_WATERMARK);
+		} catch (Exception e) {
+			LOG.warn("unable to emit watermark while closing", e);
+		}
 
-			} finally {
-				synchronized (checkpointLock) {
-					LOG.debug("Reader terminated, and exiting...");
+		cleanUp();
+	}
 
-					try {
-						this.format.closeInputFormat();
-					} catch (IOException e) {
-						getContainingTask().handleAsyncException(
-							"Caught exception from " + this.format.getClass().getName() + ".closeInputFormat() : " + e.getMessage(), e);
-					}
-					this.isSplitOpen = false;
-					this.currentSplit = null;
-					this.isRunning = false;
+	private void cleanUp() {
+		LOG.debug("cleanup, state={}", state);
 
-					checkpointLock.notifyAll();
-				}
-			}
-		}
+		RunnableWithException[] runClose = {
```

Comment:
According to @zentol , it's okay to use more Guava now, so I'd recommend to use `Closer`.

Reply:
Do you mean it's shaded version?